src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,262 @@
+  bayesmix::to_proto(state.psi, state_.mutable_psi());
+  bayesmix::to_proto(state.eta, state_.mutable_eta());
+  bayesmix::to_proto(state.lambda, state_.mutable_lambda());
+  bayesmix::to_proto(state.psi_inverse, state_.mutable_psi_inverse());
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+  out->mutable_mfa_state()->CopyFrom(state_);
+  return out;
+}
+
+void MFAHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message& hypers_) {
+  auto& hyperscast = downcast_hypers(hypers_).mfa_state();
+  hypers->mutilde = bayesmix::to_eigen(hyperscast.mutilde());
+  hypers->alpha0 = hyperscast.alpha0();
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_eta();
+    sample_mu();
+    sample_psi();
+    sample_lambda();
+  }
+}
+
+void MFAHierarchy::sample_eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  auto llt = (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+              state.lambda.transpose() * state.psi_inverse * state.lambda)
+                 .llt();
+  Eigen::MatrixXd sigma_eta(
+      llt.solve(Eigen::MatrixXd::Identity(hypers->q, hypers->q)));
+
+  if (state.eta.rows() != card) {
+    state.eta.resize(card, state.eta.cols());
+    state.eta = Eigen::MatrixXd::Zero(card, state.eta.cols());
+  }
+  for (size_t i = 0; i < card; i++) {
+    state.eta.row(i) = (stan::math::multi_normal_rng(
+        sigma_eta * (state.lambda.transpose()) * state.psi_inverse *
+            (data[i] - state.mu),
+        sigma_eta, rng));
+  }
+}
+
+void MFAHierarchy::sample_mu() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  auto llt = (hypers->phi * Eigen::MatrixXd::Identity(dim, dim) +
+              card * state.psi_inverse)
+                 .llt();
+  Eigen::MatrixXd sigma_mu = llt.solve(Eigen::MatrixXd::Identity(dim, dim));
+
+  Eigen::VectorXd sum = Eigen::VectorXd::Zero(dim);
+
+  for (size_t i = 0; i < card; i++) {
+    Eigen::VectorXd row = state.eta.row(i);
+    sum += state.lambda * row;
+  }
+  Eigen::VectorXd mumean = sigma_mu * (hypers->phi * hypers->mutilde +
+                                       state.psi_inverse * (data_sum - sum));
+
+  state.mu = stan::math::multi_normal_rng(mumean, sigma_mu, rng);
```

Comment:
I think it would be faster to write 

```
Eigen::MatrixXd sigma_inv = hypers->phi * Eigen::MatrixXd::Identity(dim, dim) +
               card * state.psi_inverse;
auto llt sigma_inv_llt = sigma_inv_llt;
Eigen::VectorXd mumean = sigma_inv_llt.solve(hypers->phi ....)
state.mu = stan::math::multi_normal_prec_rng(mumean, sigma_inv, rng);
```

Also, look at the implementation of `multi_normal_prec_rng`, if it performs a cholesky decomposition it could be worth to implement our version `multi_normal_prec_chol_rng` where we directly pass the llt factor

Reply:
Both `multi_normal_prec_rng` and `multi_normal_rng` perform a cholesky decomposition inside, so we have to use your implementation for sure to save 2*card + q inversions. Since this matrix is diagonal and very big, maybe we could even do better than this, we'll have a look at the LLT implementation in `Eigen`, but I don't think it could be automatically optimized to account for the diagonal matrix if the are not in fact DiagonalMatrix type, maybe we could just calculate the llt decomposition with coefficient-wise sqrt.

Reply:
Is your implementation called `multi_normal_prec_chol_rng`? Because in `src/utils/distributions.cc` I can find `multi_normal_prec_lpdf`for the lpdf but not the one for sampling passing the cholesky decomposition

Reply:
It's just because I we did not need it before. You can add a function `multi_normal_prec_chol_rng` in the utils/distributions file (it's just a matter of copying stan's one except for where they to the cholesky decomposition)

Reply:
Sorry I missed the information that the matrix is diagonal! Then for sure you can do (wayyyy) better: invert the diagonal element wise and sample from independent univariate normal distributions with given variance!